File: re/hacknet/题目/Decrypt.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using System.Threading;

namespace Decrypt
{
    class FileDecrypter
    {
        public string data;
        public string description;
        public string ipAddress;
        public string fileExtension;
        public ulong encryptedPassword;
        public string encryptedData;
        private const string FILE_HEADER = "#DEC_ENC";

        public FileDecrypter(string encryptedData)
        {
            this.encryptedData = encryptedData;
        }

        public ulong EncryptPassword(string password)
        {
            ulong hash1 = 5381;
            ulong hash2 = hash1;

            for (int i = 0; i < password.Length; i += 2)
            {
                hash1 = ((hash1 << 5) + hash1) ^ password[i];
                if (i == password.Length - 1)
                {
                    break;
                }
                hash2 = ((hash2 << 5) + hash2) ^ password[i + 1];
            }

            ulong hash = hash1 + hash2 * 1566083941;
            return hash % 333333333333;
        }

        public string DecryptString(string data, ulong encryptedPassword)
        {
            StringBuilder stringBuilder = new StringBuilder();
            string[] array = data.Split(new char[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
            foreach (string i in array)
            {
                ulong num = Convert.ToUInt64(i);
                char c = (char)((num - encryptedPassword - 12345678) / 19260817);
                stringBuilder.Append(c);
            }
            return stringBuilder.ToString();
        }

        public void DecryptHeader()
        {
            string[] array = encryptedData.Split(new string[] { "\r\n", "\n" }, StringSplitOptions.RemoveEmptyEntries);
            if (array.Length != 2)
            {
                throw new FormatException("文件格式不正确！");
            }
            string[] headers = array[0].Split(new string[] { "::" }, StringSplitOptions.None);
            if (headers.Length != 5 || headers[0] != FILE_HEADER)
            {
                throw new FormatException("文件格式不正确！");
            }
            ulong encryptedEmptyPassword = EncryptPassword("");
            description = DecryptString(headers[1], encryptedEmptyPassword);
            ipAddress = DecryptString(headers[2], encryptedEmptyPassword);
            fileExtension = DecryptString(headers[4], encryptedEmptyPassword);

        }

        public void DecryptBody(string password)
        {
            encryptedPassword = EncryptPassword(password);
            string[] array = encryptedData.Split(new string[] { "\r\n", "\n" }, StringSplitOptions.RemoveEmptyEntries);
            if (array.Length != 2)
            {
                throw new FormatException("文件格式不正确！");
            }
            string[] headers = array[0].Split(new string[] { "::" }, StringSplitOptions.None);
            if (headers.Length != 5 || headers[0] != FILE_HEADER)
            {
                throw new FormatException("文件格式不正确！");
            }
            if (DecryptString(headers[3], encryptedPassword) != "ENCODED")
            {
                throw new FormatException("密码错误！");
            }
            data = DecryptString(array[1], encryptedPassword);
        }

    }

    class Program
    {
        public static void Main(String[] args)
        {
            if (!File.Exists("flag.txt"))
            {
                Console.WriteLine("当前目录下不存在 flag.txt！");
                Thread.Sleep(2000);
                return;
            }

            string encryptedData = File.ReadAllText("flag.txt");
            FileDecrypter decrypter = new FileDecrypter(encryptedData);

            try
            {
                decrypter.DecryptHeader();
                Console.WriteLine("文件描述：" + decrypter.description);
                Console.WriteLine("IP地址：" + decrypter.ipAddress);
                Console.WriteLine("扩展名：" + decrypter.fileExtension);

                Console.Write("请输入密码：");
                string password = Console.ReadLine();

                decrypter.DecryptBody(password);
            }
            catch (FormatException e)
            {
                Console.WriteLine("解密失败：" + e.Message);
                Thread.Sleep(2000);
                return;
            }

            Console.WriteLine("解密成功！");
            byte[] data = Convert.FromBase64String(decrypter.data);
            File.WriteAllBytes("flag.jpg", data);
            Thread.Sleep(2000);
        }
    }
}

File: crypto/decypher/题目/Decrypt.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Base64;
import java.util.Objects;
import java.util.Scanner;
import java.util.StringJoiner;


public class Decrypt {

    public static class Decypher {

        private static final String FILE_HEADER = "#DEC_ENC";
        private static final String DELIMETER = "::";

        public static class DecFile {

            public String fileExtension;
            public String data;

            public DecFile(String fileExtension, String data) {
                this.fileExtension = fileExtension;
                this.data = data;
            }

            @Override
            public boolean equals(Object o) {
                if (this == o) {
                    return true;
                }
                if (o == null || getClass() != o.getClass()) {
                    return false;
                }
                DecFile decFile = (DecFile)o;
                return fileExtension.equals(decFile.fileExtension) &&
                        data.equals(decFile.data);
            }

            @Override
            public int hashCode() {
                return Objects.hash(fileExtension, data);
            }

            @Override
            public String toString() {
                return new StringJoiner(", ", DecFile.class.getSimpleName() + "[", "]")
                        .add("fileExtension='" + fileExtension + "'")
                        .add("data='" + data + "'")
                        .toString();
            }

        }

        private static long encryptPassword(String password) {
            long hash1 = 0x1505;
            long hash2 = hash1;

            for (int i = 0; i < password.length(); i += 2) {
                hash1 = ((hash1 << 5) + hash1) ^ password.charAt(i);
                if (i == password.length() - 1) {
                    break;
                }
                hash2 = ((hash2 << 5) + hash2) ^ password.charAt(i + 1);
            }

            long hash = hash1 + hash2 * 0x15051505;
            return (hash % 0x33333333333333L + 0x33333333333333L) % 0x33333333333333L;
        }

        private static String decryptString(String data, long encryptedPassword) {
            StringBuilder builder = new StringBuilder();
            String[] array = data.split(" +");
            for (String numString : array) {
                long num = Long.parseLong(numString);
                num = (num - encryptedPassword - 1926081719260817L) / 1145141919810L;
                builder.append((char)num);
            }
            return builder.toString();
        }

        public static DecFile decrypt(String encrypted, String password) {
            long encryptedPassword = encryptPassword(password);
            long encryptedEmptyPassword = encryptPassword("");

            String[] chunks = encrypted.split(DELIMETER, -1);
            if (chunks.length != 4 || !chunks[0].equals(FILE_HEADER)) {
                throw new IllegalArgumentException("not a valid file");
            }
            if (!"ENCODED".equals(decryptString(chunks[1], encryptedPassword))) {
                throw new IllegalArgumentException("password not correct");
            }

            String fileExtension = decryptString(chunks[2], encryptedEmptyPassword);
            String data = decryptString(chunks[3], encryptedPassword);
            return new DecFile(fileExtension, data);
        }

    }

    public static void main(String[] args) throws IOException {
        Path encryptedPath = Paths.get("flag.txt");
        if (!encryptedPath.toFile().exists()) {
            System.out.println("File not exists!");
            return;
        }

        String encryptedString = new String(Files.readAllBytes(encryptedPath));

        System.out.print("Please input password: ");
        Scanner scanner = new Scanner(System.in);
        String password = scanner.nextLine();

        try {
            Decypher.DecFile decryptedDecFile = Decypher.decrypt(encryptedString, password);
            byte[] decryptedData = Base64.getDecoder().decode(decryptedDecFile.data);
            Files.write(Paths.get("flag.txt" + decryptedDecFile.fileExtension), decryptedData);
        } catch (Exception err) {
            System.out.println("Decrypt error: " + err.getMessage());
        }
    }

}